applications/HDF5Application/python_scripts/core/file_io.py

```diff
@@ -124,32 +124,32 @@ def Get(self, model_part):
 
 class _FilenameGetterWithDirectoryInitialization(object):
 
-    def __init__(self, settings):
+    def __init__(self, settings, data_comm):
         self.filename_getter = _FilenameGetter(settings)
+        self.data_comm = data_comm
 
     def Get(self, model_part=None):
         file_name = self.filename_getter.Get(model_part)
         self._InitializeDirectory(file_name)
         return file_name
 
-    @staticmethod
-    def _InitializeDirectory(file_name):
+    def _InitializeDirectory(self, file_name):
         dirname = os.path.dirname(file_name)
         if dirname != '' and not os.path.exists(dirname):
-            if KratosMultiphysics.DataCommunicator.GetDefault().Rank() == 0:
+            if self.data_comm.Rank() == 0:
                 os.makedirs(dirname)
-            KratosMultiphysics.DataCommunicator.GetDefault().Barrier()
+            self.data_comm.Barrier()
```

Comment:
We already have problems with this barrier (@adityaghantasala  can also confirm). Even in VTK this barrier is not working as it is suppose to. The problem is as following.

For some reason in the few clusters we checked, even if the rank 0 creates the folder successfully, the folder structure is not updated in some of the computing nodes, hence after the barrier the process complains that the folder is not found when the process uses that created folder in some nodes. The same exact block is present in vtk, which is also sometimes does not help at all. This is not related to this code segment, but it might fail in some clusters. I have experienced this failing a lot. Only way is to run the process ones, and once it fails run it again so in the second time the folder structure is there.

https://github.com/KratosMultiphysics/Kratos/blob/8e95a20f9984988c33a98e77074f74edf1ce6d45/kratos/input_output/vtk_output.cpp#L234-L241

So I would not expect this barrier to solve the problem in MPI (This is evident a lot in large MPI cases)

Reply:
Do you normally define the paths (in the vtk case the `output_path`) as absolute or relative paths? I wonder if your issue comes from using relative paths

Reply:
I used the relative paths, not the absolute ones. Does it work if absolute paths are used? Only issue with absolute paths is risk of overwriting previously generated data :/

Reply:
> I used the relative paths, not the absolute ones. Does it work if absolute paths are used? 

Not sure. Im saying just to ensure that all procs point to the exact same path. Using relatives maybe the path is constructed differently? (e.g /i/am/proc0/pointing_to/vtk_output and  /i/am/proc1/pointing_to/vtk_output). Especially since partitions can be in physically different processors, and I don't know how paths are handled that way. On my experience on clusters that I've used, all have some shared filesystem and we use always absolute paths on the shared filesystem.


> Only issue with absolute paths is risk of overwriting previously generated data :/

Why would using absolute paths overwrite data? When I refer to absolute path I mean on the `folder_name`, instead of `"folder_name": "my_vtk_output"`, use `"folder_name": "/this/is/my_vtk_output"` and the "filename" can be different for each proc.

Reply:
@marcnunezc I don't see how the absolute paths should solve the problem.
I see that if the different processors would be mounted to different filesystems this would be problematic when using relative paths, but in practice (at least on the clusters we use) they are mounted to a shared filesystem. If that wouldn't be the case then our result files would appear in different places but they don't
Consider that everything works well if the folder structure exists!

It is really only the folder creation at runtime that causes the issues, both in C++ and in Python. This problem has in fact been known for years and I was already in contact with the cluster support but we couldn't figure it out. From the programming POV the code above is (still) correct to me.

The only thing I haven't tried yet is to change the `Barrier` to an explicit wait for the folder to exist
ie.e. sth lke this:
~~~
if !folder_exists && rank==0:
    / create folder
WaitForFolderToExist() // instead of Barrier
~~~

My thinking is that there is some latency in the filesystem (this is common in cluster filesystems afaik) and hence the folder is not yet "visible" on all ranks

Also pinging @mpentek as he often faces this issue too

Reply:
Ok I see! Now I recall that another thing we usually do in clusers is to create the folder structure we need in the job launching bash script `mkidr -p my_folder_1`, `mkidr -p my_folder_2`... (before Kratos is run).

Of course this is not very helpful if you have folder names that depend on the time_step or similar (that you cannot know beforehand).

In any case,  I will go on with merging this and we can continue the discussion of a possible workaround in an issue.